tests/php/ORM/ExactMatchFilterTest.php

```diff
@@ -0,0 +1,106 @@
+    {
+        parent::setUp();
+        for ($i = 0; $i < 3; $i++) {
+            Project::create([
+                'Title' => "Project $i"
+            ])->write();
+            Task::create([
+                'Title' => "Task $i",
+                'ProjectID' => $i,
+            ])->write();
+        }
+    }
+
+    /**
+     * @dataProvider provideUsePlaceholders
+     */
+    public function testUsePlaceholders(?bool $expectedID, ?bool $expectedTitle, $config, callable $fn): void
+    {
+        Config::modify()->set(ExactMatchFilter::class, 'use_placeholders_for_integer_ids', $config);
+        [$idQueryUsesPlaceholders, $titleQueryUsesPlaceholders] = $this->usesPlaceholders($fn);
+        $this->assertSame($expectedID, $idQueryUsesPlaceholders);
+        $this->assertSame($expectedTitle, $titleQueryUsesPlaceholders);
+    }
+
+    public function provideUsePlaceholders(): array
+    {
+        $ids = [1, 2, 3];
+        $taskTitles = array_map(fn($i) => "Task $i", $ids);
+        return [
+            'primary key' => [
+                false,
+                null,
+                false,
+                fn() => Task::get()->byIDs($ids)->toArray()
+            ],
+            'foriegn key' => [
+                false,
+                null,
+                false,
+                fn() => Task::get()->filter(['ProjectID' => $ids])->toArray()
+            ],
+            'regular column' => [
+                null,
+                true,
+                false,
+                fn() => Task::get()->filter(['Title' => $taskTitles])->toArray()
+            ],
+            'primary key + regular column' => [
+                false,
+                true,
+                false,
+                fn() => Task::get()->filter([
+                    'ID' => $ids,
+                    'Title' => $taskTitles
+                ])->toArray()
+            ],
+            'primary key config enabled' => [
+                true,
+                null,
+                true,
+                fn() => Task::get()->byIDs($ids)->toArray()
+            ],
+            'non int values' => [
+                true,
+                null,
+                false,
+                fn() => Task::get()->filter(['ID' => ['a', 'b', 'c']])->toArray()
+            ],
+        ];
+    }
+
+    private function usesPlaceholders(callable $fn): array
+    {
+        // force showqueries on to view executed SQL via output-buffering
+        $_REQUEST['showqueries'] = 1;
+        ob_start();
+        $fn();
+        $s = ob_get_clean();
+        unset($_REQUEST['showqueries']);
+        preg_match('#ID" IN \(([^\)]+)\)\)#', $s, $matches);
+        $idQueryUsesPlaceholders = isset($matches[1]) ? $matches[1] === '?, ?, ?' : null;
+        preg_match('#"Title" IN \(([^\)]+)\)\)#', $s, $matches);
+        $titleQueryUsesPlaceholders = isset($matches[1]) ? $matches[1] === '?, ?, ?' : null;
+        return [$idQueryUsesPlaceholders, $titleQueryUsesPlaceholders];
```

Comment:
Please use the ORM instead, e.g:
```php
// $params will contain the IDs - and we can also check the SQL itself the same way you are here
$sql = Team::get()->where(['ID in (?, ?, ?)' => [1,2,3]])->dataQuery()->sql($params);

// $params will be empty - and we can also check the SQL itself the same way you are here
$sql = Team::get()->where('ID in (1,2,3)')->dataQuery()->sql($params);
```

Reply:
I don't understand what this is supposed to do?  This method is designed to parse the sql that was generated by `Task::get()->filter(['ID' => ['a', 'b', 'c']])->toArray()` style queries.

Reply:
Calling `sql()` doesn't _write_ SQL, it shows you what the SQL _will be_ for that query. The `$params` value will have any placeholder values passed into it (it's passed by value). We don't need to execute queries to see what the SQL will look like, and using the `sql()` method is a pattern used all throughout the tests in the `DataListTest` class

So in otherwords, please remove `->toArray()` from the provided functions, and have those functions return the list. Then rewrite that method to look more like this:

```suggestion
$list = fn();
$sql = $list->dataQuery()->sql();
// we could also check the parameters as I mentioned above but since we have two possible filters, it's probably easier to continue checking the SQL like this.
preg_match('#ID" IN \(([^\)]+)\)\)#', $sql, $matches);
$idQueryUsesPlaceholders = isset($matches[1]) ? $matches[1] === '?, ?, ?' : null;
preg_match('#"Title" IN \(([^\)]+)\)\)#', $sql, $matches);
$titleQueryUsesPlaceholders = isset($matches[1]) ? $matches[1] === '?, ?, ?' : null;
return [$idQueryUsesPlaceholders, $titleQueryUsesPlaceholders];
```

Reply:
Updated